route_guide/route_guide_client.cc

```diff
@@ -23,238 +23,347 @@
+                          << data.stats.point_count() << " points\n"
+                          << "Passed " << data.stats.feature_count()
+                          << " features\n"
+                          << "Travelled " << data.stats.distance()
+                          << " meters\n"
+                          << "It took " << data.stats.elapsed_time()
+                          << " seconds" << std::endl;
+              })
+              // .finished callback is invoked when the call is finished.
+              // In this callback we just start our continuation.
+              .finished([](
+                            auto&& data,
+                            auto& continuation,
+                            auto&& response) {
+                continuation.Start(grpc::Status::OK);
+              });
   }
 
-  void RouteChat() {
-    std::thread writer;
-    Notification<Status> done;
-    client_->Call<RouteGuide, Stream<RouteNote>, Stream<RouteNote>>(
-        "RouteChat",
-        [&](auto&& call, bool ok) {
-          call->OnFinished([&](auto*, const Status& status) {
-            done.Notify(status);
-          });
-          if (!ok) {
-            call->context()->TryCancel();
-            call->Finish();
-          } else {
-            call->OnRead([](auto*, auto&& server_note) {
-              if (server_note) {
+  // RouteChat() function implements bidirectional streaming RPC. The client
+  // sends the sequence of 4 messages. Each message consists of the RouteNote
+  // data structure (you can check this data structure in the
+  // protos/route_guide.proto). The server gets this stream  of points and
+  // checks if the vector received_notes_ contains already sended points. If
+  // so, then the server prepare the stream of this points.
+  auto RouteChat() {
+    return client_->Call<
+               RouteGuide,
+               Stream<RouteNote>,
+               Stream<RouteNote>>("RouteChat")
+        | Client::Handler()
+              .context(std::thread{})
+              // Sending the request of 4 points. We will do it in another
+              // thread.
+              .ready([](auto&& writer, auto& call) {
+                writer = std::thread([&]() {
+                  const std::vector<RouteNote> notes{
+                      MakeRouteNote("First message", 0, 0),
+                      MakeRouteNote("Second message", 0, 1),
+                      MakeRouteNote("Third message", 1, 0),
+                      MakeRouteNote("Fourth message", 0, 0)};
+                  for (const RouteNote& note : notes) {
+                    std::cout << "Sending message " << note.message()
+                              << " at " << note.location().latitude()
+                              << ", " << note.location().longitude()
+                              << std::endl;
+                    call.Write(note);
+                  }
+                  call.WritesDone();
+                  writer.detach();
```

Comment:
I don't believe it's normal for `some_thread.detach()` to happen _inside_ the thread's own body. I believe it should be outside its body, on the level where `writer = std::thread(...)` happens?